### PR TITLE
Added elliptical annulus

### PR DESCRIPTION
**Pull request by siddharthlal25**
This PR is the implementation of elliptical annulus for this package.

### CODE REVIEW

**Comment by yashrsharma44** [comment]
Hey @siddharthlal25 , thanks for the PR! I think @mileslucas @giordano can suggest the logical workflow and whether it's correct, however, I can suggest some nitpicking changes that you might want to address.

File: src/Aperture/elliptical.jl [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    x::T
+    y::T
+    a::T
+    b::T
+    theta::T
+end
+
+function EllipticalAperture(x, y, a, b, theta)
+    if (b < 0.0 || a < b )
+            error("illegal ellipse parameters. Require a >= b > 0.0")
```
Comment (yashrsharma44) [inline src/Aperture/elliptical.jl]: `"Illegal..`

File: src/Aperture/elliptical.jl [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    b_in::T
+    theta_in::T
+    a_out::T
+    b_out::T
+    theta_out::T
+end
+
+function EllipticalAnnulus(x, y, a_in, b_in, theta_in, a_out, b_out, theta_out)
+    if (b_in < 0.0 || a_in < b_in || b_out < 0.0 || a_out < b_out)
+            error("illegal ellipse parameters. Require a_in >= b_in > 0.0, a_out >= b_out > 0.0")
```
Comment (yashrsharma44) [inline src/Aperture/elliptical.jl]: `"Illegal"`

File: src/Aperture/overlap.jl [hunk 3]
```diff
@@ -154,3 +154,105 @@ function area_arc(x0, y0, x1, y1, r)
...
+        pxmax = pxmin + dx
+
+        if pxmax > bxmin && pxmin < bxmax
+            for j in 1:ny
+                pymin = ymin + (j - 1) * dy
+                pycen = pymin + 0.5dy
+                pymax = pymin + dy
+                if pymax > bymin && pymin < bymax
+
+                    #4 flags for four different ends of the pixel
```
Comment (yashrsharma44) [inline src/Aperture/overlap.jl]: Maybe you can add some descriptive variable names for the `flag` variable?

File: docs/src/apertures/apertures.md [hunk 4]
```diff
@@ -24,7 +24,12 @@ CircularAnnulus
 
 ## Elliptical Apertures
 
-These apertures are parametrized by the semi-major axis `a` and semi-minor axis `b`.
+These apertures are parametrized by the semi-major axis `a`, semi-minor axis `b` and rotation angle in degrees clockwise from the positive x-axis `theta`
```
Comment (yashrsharma44) [inline docs/src/apertures/apertures.md]:
```suggestion
These apertures are parametrized by the semi-major axis `a`, semi-minor axis `b` and rotation angle in degrees clockwise from the positive x-axis `theta`.
```

File: test/overlap.jl [hunk 5]
```diff
@@ -63,3 +64,9 @@ end
 
 
 end
+
```
Comment (yashrsharma44) [inline test/overlap.jl]: I can see that some tests for the remaining functions are missing, it would be great if you can them in this PR!

File: src/Aperture/elliptical.jl (hunk 1 above)
Comment (mileslucas) [reply src/Aperture/elliptical.jl]: I don't like the term "illegal". Perhaps use "invalid" instead.

File: src/Aperture/elliptical.jl (hunk 2 above)
Comment (mileslucas) [reply src/Aperture/elliptical.jl]: See above

**Comment by mileslucas** [comment]
Is this intended to build on top of #4 or to be built in parallel? The overlap code for the annuli should not be different than the aperture.

**Comment by siddharthlal25** [comment]
> Is this intended to build on top of #4 or to be built in parallel? The overlap code for the annuli should not be different than the aperture.

The overlap would be the same for annuli and aperture but the bbox for annuli would be built upon bbox of aperture so I thought to make annuli separately. What do you suggest, should I merge the #8 in #4?

**Comment by mileslucas** [comment]
Ah let me rephrase my statement:

Does this PR depend on #4 being merged to be valid?

As for the bboxes, the box for any annular version is the same as the aperture version but using the outer parameters. EG circular annuli are the same as a circular aperture with radius `r_out`.

**Comment by mileslucas** [comment]
Please rebase off of #4 and continue developing.

**Comment by yashrsharma44** [comment]
@siddharthlal25 Here is a good writeup for rebasing in case you are interested - [Link](https://github.com/k88hudson/git-flight-rules#rebasing-and-merging)

**Comment by yashrsharma44** [comment]
@siddharthlal25 Code wise looks good to me!

File: src/Aperture/elliptical.jl [hunk 6]
```diff
@@ -94,3 +94,57 @@ function mask(e::EllipticalAperture; method = :center)
...
+An elliptical annulus with semi-major axis `a`, semi-minor axis `b`, angle `θ` of inner ellipse ,scaling factor `factor` to generate the outer ellipse from inner ellipse.
+`a` and `b` must be ≥ 0, factor ≥ 1, `θ` is measured in degrees counter-clockwise the standard x-axis.
+
+# Examples
+```jldoctest
+julia> ap = EllipticalAnnulus(0, 0, 10, 5, 45, 2)
+EllipticalAnnulus(0, 0, a=10, b=5, θ=25°, factor=2)
+
+```
+"""
```
Comment (mileslucas) [inline src/Aperture/elliptical.jl]: Actually, I think the better way to parametrize this is
`EllipticalAnnulus(x, y, a_in, a_out, b_out, θ)`
and just calculate `b_in = a_in / a_out * b_out`.

This matches `photutils` so hopefully that removes a possible pain-point for future users.

File: src/Aperture/elliptical.jl [hunk 7]
```diff
@@ -94,3 +94,57 @@ function mask(e::EllipticalAperture; method = :center)
...
+        factor < 1 && error("Invalid scaling factor=$b. factor must be greater than or equal to 1")
+        new{T}(x, y, a, b, mod(theta, 360), factor)
+    end
+end
+
+EllipticalAnnulus(x, y, a, b, theta, factor) = EllipticalAnnulus(promote(x, y, a, b, theta, factor)...)
+EllipticalAnnulus(center::AbstractVector, a, b, theta, factor) = EllipticalAnnulus(center..., a, b, theta, factor)
+
+function Base.show(io::IO, e::EllipticalAnnulus)
+    print(io, "EllipticalAnnulus($(e.x), $(e.y), a=$(e.a), b=$(e.b), θ=$(e.theta)°, factor=$(e.factor))")
```
Comment (mileslucas) [inline src/Aperture/elliptical.jl]: switch to `a_in=$(e.a_in), a_out=$(e.a_out), b_in=$(e.b_in), b_out=$(e.b_out)`

File: test/photometry.jl [hunk 8]
```diff
@@ -9,17 +9,20 @@ PARAMS = [
...
-    EllipticalAperture
+    EllipticalAperture,
+    EllipticalAnnulus
 ]
 
 ELL_PARAMS = [
-    (3, 3, 0)
+    (3, 3, 0),
+    (3, 3, 0, 5/3)
 ]
```
Comment (mileslucas) [inline test/photometry.jl]: Now that exact method works, just lump into the existing test code (you'll see where it should go).

File: src/Aperture/elliptical.jl [hunk 9]
```diff
@@ -94,3 +94,57 @@ function mask(e::EllipticalAperture; method = :center)
...
+end
+
+function bbox(e::EllipticalAnnulus)
+    xmin, xmax, ymin, ymax = bbox(EllipticalAperture(e.x, e.y, e.a * e.factor, e.b * e.factor, e.theta))
```
Comment (mileslucas) [inline src/Aperture/elliptical.jl]: I would like to see this without having to spend the allocation on creating another type. The actual code might be repetitive but it'll be more memory-efficient (and marginally faster) without the allocate.

**Comment by mileslucas** [comment]
Oh, another comment: add an annulus to the `docs/src/apertures/examples.md` first plot for demonstration.

Thanks for all the effort for this!

File: src/Aperture/elliptical.jl [hunk 10]
```diff
@@ -98,3 +98,95 @@ function mask(e::EllipticalAperture; method = :center)
...
+    b_in::T
+    a_out::T
+    b_out::T
+    theta::T
+
+    function EllipticalAnnulus(x::T, y::T, a_in::T, b_in::T, a_out::T, b_out::T, theta::T) where T <: Number
+        a_in < 0 && error("Invalid axis a_in=$a_in. a_in must be greater than or equal to 0")
+        b_out < 0 && error("Invalid axis b_out=$b_out. b_out must be greater than or equal to 0")
+        a_out < 0 && error("Invalid axis a_out=$a_out. a_out must be greater than or equal to 0")
+        a_in > a_out && error("Invalid axis a_in=$a_in. a_in must be less than or equal to a_out")
```
Comment (mileslucas) [inline src/Aperture/elliptical.jl]: we can simplify this
```julia
0 ≤ a_in ≤ a_out || error("...")
0 ≤ b_in ≤ b_out || error("...")
```

File: src/Aperture/elliptical.jl [hunk 11]
```diff
@@ -98,3 +98,95 @@ function mask(e::EllipticalAperture; method = :center)
...
 end
+
+
+#######################################################
+
+"""
+    EllipticalAnnulus(x, y, a_in, a_out, b_out, θ)
+    EllipticalAnnulus([x, y], a_in, a_out, b_out, θ)
+An elliptical annulus with inner semi-major axis `a_in` outer semi-major axis `a_out`, outer semi-minor axis `b_out`, angle `θ` of inner ellipse.
+`a_out` ≥ `a_in` ≥ 0 and `b` must be ≥ 0, `θ` is measured in degrees counter-clockwise the standard x-axis.
```
Comment (mileslucas) [inline src/Aperture/elliptical.jl]: **position** angle

File: src/Aperture/plotting.jl [hunk 12]
```diff
@@ -51,3 +51,27 @@ end
...
+
+    # inner ring
+    @series begin
+        x = @. e.x + e.a * cos(t) * cosd(e.theta) - e.b * sin(t) * sind(e.theta) + 0.5
+        y = @. e.y + e.a * cos(t) * sind(e.theta) + e.b * sin(t) * cosd(e.theta) + 0.5
+        x, y
+    end
+
+
+end
```
Comment (mileslucas) [inline src/Aperture/plotting.jl]: needs rewritten

**Comment by mileslucas** [comment]
@JuliaRegistrator register
